### DISCUSSION TITLE
DAG state is Success even in case of intermediate Task Failure for TriggerRule.ALL_DONE

### DISCUSSION THREAD

**Discussion by pateash** [apache/airflow#24190] [General]
**Description**
I am not sure if this is BUG, that's why creating as a Feature Request

**Minimal code to reproduce**

    args = {
    'owner': 'airflow',
    }

    with DAG(
        dag_id='dag_state_test',
        default_args=args,
        schedule_interval='0 0 * * *',
        start_date=days_ago(2),
    ) as dag:

    run_this = BashOperator(
        task_id='middle',
        bash_command='echo "middle" && exit 1',
    )

    startOperator = BashOperator(
        task_id='start',
        bash_command='echo "start"',
    )

    endOperator = BashOperator(
        task_id='end',
        bash_command='echo "end"',
        trigger_rule=TriggerRule.ALL_DONE  # so, this runs always no matter what happens in middle.
    )
    startOperator >> run_this >> endOperator

**Description** 
In our case we have following pipelines

    DataProcCreateCluster() -> Spark Jobs -> DataProcDeleteCluster()

**DataProcDeleteCluster()** has to run always no matter what happens in between ( TriggerRule.ALL_DONE used), 
as it looks like the DAG state is inherited only from state of last task.

![image](https://user-images.githubusercontent.com/16856802/115519682-0ca04400-a2a7-11eb-9c3c-9611317e141b.png)

**Use case / motivation**
We should be able to set DAG state using rules like ALL_SUCCESS ( only set success if all tasks are success)

**Are you willing to submit a PR?**
YES

**Related Issues**
NONE

**Comment by pateash** [comment]
### Workaround
As the task only look into parents for the state, we can create a **DummyOperator** which has dependency overall 



    ......

    dummy_end = DummyOperator(
        task_id='dummy_end',
        trigger_rule=TriggerRule.NONE_FAILED_OR_SKIPPED

    )

    .........


    startOperator >> dummy_end
    run_this >> dummy_end
    endOperator >> dummy_end

![image](https://user-images.githubusercontent.com/16856802/115522012-58ec8380-a2a9-11eb-9fc2-0cd94cb54b0e.png)

**Comment by eladkal** [comment]
I'm sorry but I don't understand what is the feature request here?
`TriggerRule.ALL_DONE` works fine with `Success`, `Failed` and `Skipped`.

```
from airflow import DAG
from airflow.operators.python import PythonOperator
from airflow.operators.dummy import DummyOperator
from airflow.utils.trigger_rule import TriggerRule
from airflow.exceptions import AirflowSkipException, AirflowException
import datetime


def func():
    raise AirflowSkipException()


def func2():
    raise AirflowException()


args = {
    'owner': 'airflow',
}

with DAG(
    dag_id='15468',
    default_args=args,
    schedule_interval=None,
    start_date=datetime.datetime(2020, 3, 4),
) as dag:

    skip_op = PythonOperator(
        task_id='skip_task',
        python_callable=func,
    )

    fail_op = PythonOperator(
        task_id='fail_task',
        python_callable=func2,
    )

    success_op = DummyOperator(
        task_id='success_task',
    )
    end_op = DummyOperator(
        task_id='end',
        trigger_rule=TriggerRule.ALL_DONE  # so, this runs always no matter what happens in middle.
    )
    [skip_op, fail_op, success_op] >> end_op
```


![Screen Shot 2021-06-11 at 22 24 49](https://user-images.githubusercontent.com/45845474/121739149-e14a1080-cb03-11eb-9052-a21d49f58591.png)

**Comment by pateash** [comment]
> I'm sorry but I don't understand what is the feature request here?
> `TriggerRule.ALL_DONE` works fine with `Success`, `Failed` and `Skipped`.
> 
> ```
> from airflow import DAG
> from airflow.operators.python import PythonOperator
> from airflow.operators.dummy import DummyOperator
> from airflow.utils.trigger_rule import TriggerRule
> from airflow.exceptions import AirflowSkipException, AirflowException
> import datetime
> 
> 
> def func():
>     raise AirflowSkipException()
> 
> 
> def func2():
>     raise AirflowException()
> 
> 
> args = {
>     'owner': 'airflow',
> }
> 
> with DAG(
>     dag_id='15468',
>     default_args=args,
>     schedule_interval=None,
>     start_date=datetime.datetime(2020, 3, 4),
> ) as dag:
> 
>     skip_op = PythonOperator(
>         task_id='skip_task',
>         python_callable=func,
>     )
> 
>     fail_op = PythonOperator(
>         task_id='fail_task',
>         python_callable=func2,
>     )
> 
>     success_op = DummyOperator(
>         task_id='success_task',
>     )
>     end_op = DummyOperator(
>         task_id='end',
>         trigger_rule=TriggerRule.ALL_DONE  # so, this runs always no matter what happens in middle.
>     )
>     [skip_op, fail_op, success_op] >> end_op
> ```
> 
> ![Screen Shot 2021-06-11 at 22 24 49](https://user-images.githubusercontent.com/45845474/121739149-e14a1080-cb03-11eb-9052-a21d49f58591.png)

sure @eladkal , In your example 
the DAG state will be **success** even though one of the intermediate tasks has been failed,
because the DAG state only depends on the state of the last task.
which may not be a requirement in many of the cases and should be configurable.

**Comment by eladkal** [comment]
So remove end_op from the code and it will work as you expect, isn't it?

**Comment by pateash** [comment]
> So remove end_op from the code and it will work as you expect, isn't it?

yeah, but what if you wanted end_op to run.
ex.

you are created a GCP Dataproc or AWS EMR Cluster and running bunch of tasks, your last task is Delete cluster.
if any of the tasks failed, you still wanted to delete your cluster but the status of dag must be failed.

**Comment by eladkal** [comment]
I have similar use case but I don't cause the DAG itself to fail if EMR failed. Like you I run a task task to terminate the machine.
Am I correct that what you are asking is to change the way that Airflow consider DAG runs statuses (success/failure) by giving the user the option to set it according to his own logic?

**Comment by pateash** [comment]
> I have similar use case but I don't cause the DAG itself to fail if EMR failed. Like you I run a task task to terminate the machine.
> Am I correct that what you are asking is to change the way that Airflow consider DAG runs statuses (success/failure) by giving the user the option to set it according to his own logic?

yeah, that's correct.

**Comment by pateash** [comment]
@eladkal @potiuk @ashb @kaxil @mik-laj 

Does this requirement looks legit?
I would like to work on it, if this looks ok.

**Comment by pateash** [comment]
> I have similar use case but I don't cause the DAG itself to fail if EMR failed. Like you I run a task task to terminate the machine. Am I correct that what you are asking is to change the way that Airflow consider DAG runs statuses (success/failure) by giving the user the option to set it according to his own logic?

@eladkal  In your case, if you consider dag to be successful even though intermediate jobs have failed,
How are you monitoring your jobs and Isn't the dag should fail ideally in your case?

**Comment by kaxil** [comment]
Like @eladkal mentioned - I have had similar requirements like him and I would have `on_failure_callbacks` notifying that my task failed.

For your case, you can write a CustomOperator which you have at the end that checks if your EMR task failed or not and fail this task itself, thereby the DagRun will also be failed as it is the last task.


---

Your feature request while legit need more thinking through:
- What happens when there are more than one terminal tasks
- Do we want to allow users to set which task_id need to succeed for DagRun status

**Comment by potiuk** [comment]
this is as expected We implemented some solutions for that for AIP-47 system tests. There is a simple pattern of adding a "watcher" task to monitor tasks that should fail whole dag in case any of upstream tasks failed and I think this answers all the questions. See https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-47+New+design+of+Airflow+System+Tests for details. 

Converting it into discussion in case more discussion is needed.
